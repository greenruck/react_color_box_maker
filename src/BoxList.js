import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
    const [boxes, setboxes] = useState([]);
    const add = boxObj => {
        setboxes(boxes => [...boxes, boxObj]);
    };
    const remove = id => {
        setboxes(boxes => boxes.filter(box => box.id !== id));
    };
    const boxComponents = boxes.map(box => (
        <Box
            key = {box.id}
            id = {box.id}
            color={box.color}
            width = {box.width}
            height = {box.height}
            handleRemove = {remove}
        />
    ));

    return (
        <div>
            <NewBoxForm createBox = {add} />
            {boxComponents}
        </div>
    );
}

export default BoxList;