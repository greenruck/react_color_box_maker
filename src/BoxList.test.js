import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "4", width = "4", color = "green") {
    const heightInput = boxList.getByLabelText("Height");
    const widthInput = boxList.getByLabelText("Width");
    const colorInput = boxList.getByLabelText("Color");

    fireEvent.change(colorInput, { target: { value: color } });
    fireEvent.change( widthInput, {target: { value: width } });
    fireEvent.change( heightInput, {target: { value: height } });

    // const button = boxList.getBytext("Add a box");
    fireEvent.click(boxList.getBytext("Add a box"));
}

it("renders without crashing", function() {
    render(<BoxList />);
  });
  
it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add new boxes", function() {
    const boxList = render(<BoxList />);

    expect(boxList.queryByText("Remove Box")).not.toBeInTheDocument();

    addBox(boxList);

    // const removeButton = boxList.getByText("Remove Box");
    expect(boxList.getByText("Remove Box")).toBeInTheDocument();
    expect(boxList.getByText("Remove Box").previousSibling).toHaveStyle(`
      width: 4em;
      height: 4em;
      background-color: green;
    `);

    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);
});

it("can remove a box", function() {
    const boxList = render(<BoxList />);
    addBox(boxList);
  
    // const removeButton = boxList.getByText("Remove Box");
  
    fireEvent.click(boxList.getByText("Remove Box"));
    expect(boxList.getByText("Remove Box")).not.toBeInTheDocument();
  });
  