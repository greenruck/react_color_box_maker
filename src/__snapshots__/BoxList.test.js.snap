// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <div
          class="height"
        >
          <label
            for="height"
          >
            Height
          </label>
          <input
            id="height"
            name="height"
            type="text"
            value=""
          />
        </div>
        <div
          class="width"
        >
          <label
            for="width"
          >
            Width
          </label>
          <input
            id="width"
            name="width"
            type="text"
            value=""
          />
        </div>
        <div
          class="color"
        >
          <label
            for="color"
          >
            Color
          </label>
          <input
            id="color"
            name="color"
            type="text"
            value=""
          />
        </div>
        <button
          id="newBoxButton"
        >
          Add a box
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
